File: 0239-sliding-window-maximum/0239-sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        multiset<int> mul;
        int i=0;
        int j=0;
        int n=nums.size();
        vector<int> arr;
        int p=k;
        while(k--)mul.insert(nums[j++]);
        //cout<<j<<endl;
        while(j<n){
            auto it=mul.rbegin();
            //cout<<*it<<" ";
            arr.push_back(*it);
            mul.erase(mul.find(nums[i++]));
            mul.insert(nums[j++]);
        }
        arr.push_back(*mul.rbegin());
        //cout<<*mul.rbegin()<<" ";
        return arr;
    }
};

File: 0239-sliding-window-maximum/0239-sliding-window-maximum.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        int n=nums.length;
        int size=n-(k-1);
        int[] ans=new int[size];
        int index=0;
        TreeMap<Integer,Integer> map=new TreeMap<>();
        for(int i=0;i<n;i++){
            if(i<k){
                map.put(nums[i],map.getOrDefault(nums[i],0)+1);
            }else{
                map.put(nums[i],map.getOrDefault(nums[i],0)+1);
                int val=map.get(nums[i-k]);
                if(val==1)map.remove(nums[i-k]);
                else map.put(nums[i-k],val-1);
            }
            if(i>=k-1)
            ans[index++]=map.lastKey();
        }return ans;
    }
}